builder/dockerfile/copy_windows.go

```diff
@@ -1,23 +1,99 @@
 package dockerfile
 
 import (
+	"bytes"
 	"errors"
+	"fmt"
+	"os"
 	"path/filepath"
 	"strings"
 
+	"github.com/Microsoft/go-winio"
 	"github.com/docker/docker/pkg/idtools"
+	"github.com/docker/docker/pkg/reexec"
+	"github.com/docker/docker/pkg/system"
+	"golang.org/x/sys/windows"
 )
 
 var pathBlacklist = map[string]bool{
 	"c:\\":        true,
 	"c:\\windows": true,
 }
 
-func fixPermissions(source, destination string, rootIDs idtools.IDPair, overrideSkip bool) error {
-	// chown is not supported on Windows
+func init() {
+	reexec.Register("windows-fix-permissions", fixPermissionsReexec)
+}
+
+func fixPermissions(source, destination string, identity idtools.Identity, _ bool) error {
+	if identity.SID == "" {
+		return nil
+	}
+
+	cmd := reexec.Command("windows-fix-permissions", source, destination, identity.SID)
+	output := bytes.NewBuffer(nil)
+	cmd.Stdout = output
+	cmd.Stderr = output
```

Comment:
The output is separate because it's not shown except in the case of an error.

Reply:
I don't understand your response.

I'm suggesting the 2 lines I have above are equivalent to the 12 lines (33-45) you have now.

Reply:
Yes, I realized that after I had written the comment (I misunderstood what it did), but I couldn't edit the comment :)

Reply:
I'm not convinced that CombinedOutput() is the correct approach. I need to set the output buffer, prior to executing the reexec command, not get the output buffer that the command used.

Reply:
[Cmd.CombinedOutput](https://golang.org/pkg/os/exec/#Cmd.CombinedOutput)

> CombinedOutput runs the command and returns its combined standard output and standard error.

You don't actually care about the output buffer, right? What you care about is getting the stdout and stderr for the command that run. You could do that by setting a buffer yourself and reading it. Or you can use the stdlib that already provides this functionality.

Here's the implementation: https://golang.org/src/os/exec/exec.go#L496
It should look familiar because it's basically the same thing as you have here.